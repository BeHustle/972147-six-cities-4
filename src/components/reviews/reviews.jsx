import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import AddReview from '../add-review/add-review.jsx';
import Review from '../review/review.jsx';

const Reviews = ({reviews, offerId}) => {
  const filteredReviews = reviews.filter((review) => review.offerId === offerId);
  return <section className="property__reviews reviews">
    <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{filteredReviews.length}</span></h2>
    <ul className="reviews__list">
      {filteredReviews.map((review) =>
        <Review key={review.id} review={review}/>
      )}
    </ul>
    <AddReview/>
  </section>;
};

Reviews.propTypes = {
  offerId: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.exact({
    id: PropTypes.number.isRequired,
    offerId: PropTypes.number.isRequired,
    author: PropTypes.exact({
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string
    }),
    text: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    date: PropTypes.instanceOf(Date)
  }))
};

const mapStateToProps = (state) => ({
  reviews: state.reviews
});

export default connect(mapStateToProps)(Reviews);
