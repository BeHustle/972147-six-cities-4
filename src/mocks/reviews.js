export const reviews = [
  {
    id: 1,
    offerId: 1,
    author: {
      name: `Max`,
      avatar: `img/avatar-angelina.jpg`
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit. Minima, quis!`,
    rating: 5,
    date: new Date(`2019-12-17`)
  },
  {
    id: 2,
    offerId: 1,
    author: {
      name: `Alex`,
      avatar: `img/avatar-max.jpg`
    },
    text: `Lorem ipsum dolor sit amet!`,
    rating: 3,
    date: new Date(`2020-07-17`)
  },
  {
    id: 3,
    offerId: 3,
    author: {
      name: `William`,
      avatar: ``
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit. At ipsa labore libero nam sunt temporibus?`,
    rating: 1,
    date: new Date(`2018-03-25`)
  },
  {
    id: 4,
    offerId: 2,
    author: {
      name: `Lucy`,
      avatar: `img/avatar-angelina.jpg`
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
    rating: 4,
    date: new Date(`2020-01-13`)
  },
];
