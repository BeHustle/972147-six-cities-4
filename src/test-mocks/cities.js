export const cities = [
  {
    id: 1,
    name: `Amsterdam`,
    coordinates: [52.38333, 4.9],
    zoom: 10
  },
  {
    id: 2,
    name: `Paris`,
    coordinates: [48.85875194, 2.33926391],
    zoom: 10
  },
  {
    id: 3,
    name: `Cologne`,
    coordinates: [50.938361, 6.959974],
    zoom: 10
  },
  {
    id: 4,
    name: `Brussels`,
    coordinates: [50.8465573, 4.351697],
    zoom: 10
  },
  {
    id: 5,
    name: `Hamburg`,
    coordinates: [53.550341, 10.000654],
    zoom: 10
  },
  {
    id: 6,
    name: `Dusseldorf`,
    coordinates: [51.2254018, 6.7763137],
    zoom: 10
  },
];
