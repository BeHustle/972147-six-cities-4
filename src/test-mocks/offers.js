export const offers = [
  {
    id: 1,
    name: `Beautiful & luxurious apartment at great locatccion`,
    price: 132,
    image: `img/apartment-01.jpg`,
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/apartment-01.jpg`,
    ],
    type: `hotel`,
    isPremium: true,
    inBookmarks: true,
    rating: 4.1,
    rooms: `3 Bedrooms`,
    guests: `Max 4 adults`,
    facilities: [`Wi-Fi`,
      `Washing machine`,
      `Towels`,
      `Heating`,
      `Coffee machine`,
      `Baby seat`,
      `Kitchen`,
      `Dishwasher`,
      `Cabel TV`,
      `Fridge`,
    ],
    author: {
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`,
      isSuper: true,
    },
    text: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.`,
    ],
    coordinates: [52.3909553943508, 4.85309666406198],
    cityId: 1,
  },
  {
    id: 2,
    name: `Wood and stone place`,
    price: 254,
    image: `img/apartment-02.jpg`,
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/apartment-01.jpg`,
    ],
    type: `apartment`,
    isPremium: false,
    inBookmarks: false,
    coordinates: [52.369553943508, 4.85309666406198],
    rating: 5,
    rooms: `1 Bedrooms`,
    guests: `Max 2 adults`,
    facilities: [`Wi-Fi`,
      `Washing machine`,
      `Towels`,
    ],
    author: {
      name: `Max`,
      avatar: `img/avatar-max.jpg`,
      isSuper: true,
    },
    text: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.`,
    ],
    cityId: 1,
  },
  {
    id: 3,
    name: `Canal View Prinsengracht`,
    price: 178,
    image: `img/apartment-03.jpg`,
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/apartment-01.jpg`,
    ],
    type: `room`,
    isPremium: true,
    inBookmarks: true,
    rating: 3,
    coordinates: [48.85875194, 2.33926391],
    rooms: `2 Bedrooms`,
    guests: `Max 2 adults and 1 children`,
    facilities: [`Wi-Fi`,
      `Heating`,
      `Coffee machine`,
      `Baby seat`,
      `Kitchen`,
    ],
    author: {
      name: `Alex`,
      avatar: ``,
      isSuper: true,
    },
    text: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.`,
    ],
    cityId: 2,
  },
  {
    id: 4,
    name: `Nice, cozy, warm big bed apartment`,
    price: 995,
    image: `img/room.jpg`,
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/apartment-01.jpg`,
    ],
    type: `house`,
    isPremium: true,
    inBookmarks: false,
    coordinates: [50.938361, 6.959974],
    rating: 1,
    rooms: `5 Bedrooms`,
    guests: `Max 3 adults and 2 children`,
    facilities: [`Wi-Fi`,
      `Kitchen`,
      `Dishwasher`,
      `Cabel TV`,
      `Fridge`,
    ],
    author: {
      name: `Steven`,
      avatar: ``,
      isSuper: true,
    },
    text: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.`,
    ],
    cityId: 3,
  }];

export const serverOffers = [
  {
    "bedrooms": 3,
    "city": {
      "location": {
        "latitude": 52.370216,
        "longitude": 4.895168,
        "zoom": 10
      },
      "name": `Amsterdam`
    },
    "description": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    "goods": [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    "host": {
      "avatar_url": `img/1.png`,
      "id": 3,
      "is_pro": true,
      "name": `Angelina`
    },
    "id": 1,
    "images": [`img/1.png`, `img/2.png`],
    "is_favorite": false,
    "is_premium": false,
    "location": {
      "latitude": 52.35514938496378,
      "longitude": 4.673877537499948,
      "zoom": 8
    },
    "max_adults": 4,
    "preview_image": `img/1.png`,
    "price": 120,
    "rating": 4.8,
    "title": `Beautiful & luxurious studio at great location`,
    "type": `apartment`
  }
];
