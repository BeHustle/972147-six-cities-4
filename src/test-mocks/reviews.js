export const reviews = [
  {
    id: 1,
    offerId: 1,
    author: {
      name: `Max`,
      avatar: `img/avatar-angelina.jpg`
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit. Minima, quis!`,
    rating: 5,
    date: new Date(`2019-12-17`)
  },
  {
    id: 2,
    offerId: 1,
    author: {
      name: `Alex`,
      avatar: `img/avatar-max.jpg`
    },
    text: `Lorem ipsum dolor sit amet!`,
    rating: 3,
    date: new Date(`2020-07-17`)
  },
  {
    id: 3,
    offerId: 3,
    author: {
      name: `William`,
      avatar: ``
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit. At ipsa labore libero nam sunt temporibus?`,
    rating: 1,
    date: new Date(`2018-03-25`)
  },
  {
    id: 4,
    offerId: 2,
    author: {
      name: `Lucy`,
      avatar: `img/avatar-angelina.jpg`
    },
    text: `Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
    rating: 4,
    date: new Date(`2020-01-13`)
  },
];

export const serverReviews = [
  {
    "comment": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    "date": `2019-05-08T14:13:56.569Z`,
    "id": 1,
    "rating": 4,
    "user": {
      "avatar_url": `img/1.png`,
      "id": 4,
      "is_pro": false,
      "name": `Max`
    }
  }
];
