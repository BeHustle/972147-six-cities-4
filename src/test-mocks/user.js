export const userInfo = {
  img: `user.png`,
  email: `user@gmail.com`,
  name: `User`,
  isPro: true
};

export const serverUserInfo = {
  "avatar_url": `img/1.png`,
  "email": `Oliver.conner@gmail.com`,
  "id": 1,
  "is_pro": false,
  "name": `Oliver.conner`
};

export const postUserInfo = {
  email: `user@mail.ru`,
  password: `12345678`
};
