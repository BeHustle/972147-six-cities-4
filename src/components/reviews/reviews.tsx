import * as React from 'react';
import {connect} from 'react-redux';
import {AuthStatus} from '../../constants';
import {Operation as DataOperation} from '../../reducer/data/data.reducer';
import {getReviews} from '../../reducer/data/data.selectors';
import {getAuthStatus} from '../../reducer/user/user.selectors';
import AddReview from '../add-review/add-review';
import Review from '../review/review';
import {ReviewInterface} from "../../types";

interface Props {
  offerId: number;
  onReviewsMount: (offerId: number) => void;
  reviews: Array<ReviewInterface>;
  authStatus: string;
}

class Reviews extends React.PureComponent<Props, {}> {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onReviewsMount, offerId} = this.props;
    onReviewsMount(offerId);
  }

  render() {
    const {reviews, authStatus, offerId} = this.props;
    return <section className="property__reviews reviews">
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
      <ul className="reviews__list">
        {reviews.map((review) =>
          <Review key={review.id} review={review}/>,
        )}
      </ul>
      {authStatus === AuthStatus.AUTH && <AddReview offerId={offerId} />}
    </section>;
  }
}

const mapDispatchToProps = (dispatch) => ({
  onReviewsMount(offerId) {
    dispatch(DataOperation.loadReviews(offerId));
  },
});

const mapStateToProps = (state) => ({
  reviews: getReviews(state),
  authStatus: getAuthStatus(state)
});

export {Reviews};

export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
