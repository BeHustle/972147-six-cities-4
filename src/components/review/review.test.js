import * as React from 'react';
import renderer from 'react-test-renderer';
import Review from './review.tsx';
import {reviews} from '../../test-mocks/reviews';

it(`Render review form`, () => {
  const tree = renderer
    .create(<Review review={reviews[0]} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
