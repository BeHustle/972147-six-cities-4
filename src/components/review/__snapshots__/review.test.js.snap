// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render review form 1`] = `
<li
  className="reviews__item"
>
  <div
    className="reviews__user user"
  >
    <div
      className="reviews__avatar-wrapper user__avatar-wrapper"
    >
      <img
        alt="Max"
        className="reviews__avatar user__avatar"
        height="54"
        src="img/avatar-angelina.jpg"
        width="54"
      />
    </div>
    <span
      className="reviews__user-name"
    >
      Max
    </span>
  </div>
  <div
    className="reviews__info"
  >
    <div
      className="reviews__rating rating"
    >
      <div
        className="reviews__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <p
      className="reviews__text"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Minima, quis!
    </p>
    <time
      className="reviews__time"
      dateTime="December 17th 2019-12-17"
    >
      December 2019
    </time>
  </div>
</li>
`;
